packages/desktop/src/renderer/components/hooks/useInput.ts

```diff
@@ -0,0 +1,11 @@
+import { useEngineStore } from '@renderer/engine'
+
+export const useInput = (id: string) => {
+  const inputs = useEngineStore((state) => state.inputs)
+  return inputs[id]
+}
+
+export const useInputsWithNode = (nodeId: string) => {
+  const inputs = useEngineStore((state) => state.inputs)
+  return Object.values(inputs).filter((input) => input.targetNodeIds.includes(nodeId))
+}
```

Comment:
Similar issue here, but a bit more complex. Your filtered list of inputs will always return a new array, thus triggering a re-render. `Object.values` and `Array.filter` always generates a new list, and so will always cause a re-render. Zustand provides a hook to fix this, `useShallow`, which performs a shallow equality check on your filtered output to see if it actually changed.

```ts
export const useInputsWithNode = (nodeId: string) => {
  return useEngineStore(
    useShallow((state) => 
      Object.values(state.inputs).filter((input) => input.targetNodeIds.includes(nodeId))
    )
  );
};
```